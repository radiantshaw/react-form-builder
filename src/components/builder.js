import React, { forwardRef, useContext } from "react";

import deriveName from "../utils/derive-name";

const FormBuilderContext = React.createContext();

function useDerivedName(name) {
  if (!name) return null;

  return deriveName([
    ...useContext(FormBuilderContext), name
  ]);
}

let builder = {};

builder.form = forwardRef(function(props, ref) {
  return (
    <form ref={ref}>
      <FormBuilderContext.Provider value={[props.name]}>
        { props.children }
      </FormBuilderContext.Provider>
    </form>
  );
});

builder.fields = function(props) {
  if (props.name) {
    var formContext =
      props.formName ? [props.formName] : [...useContext(FormBuilderContext)];
  } else {
    return props.children || null;
  }

  return (
    <FormBuilderContext.Provider
      value={[...formContext, props.name]}>
      { props.children }
    </FormBuilderContext.Provider>
  );
}

builder.input = forwardRef(function(props, ref) {
  return (
    <input ref={ref}
      name={useDerivedName(props.name)}
    />
  );
});

builder.select = forwardRef(function(props, ref) {
  return (
    <select ref={ref}
      name={useDerivedName(props.name)}>
      { props.children }
    </select>
  );
});

builder.textarea = forwardRef(function(props, ref) {
  return (
    <textarea ref={ref}
      name={useDerivedName(props.name)}
    />
  );
});

export default builder;
