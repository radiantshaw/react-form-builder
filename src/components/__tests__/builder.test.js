import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import pretty from "pretty";

import builder from "../builder";

let container = null;

beforeEach(function() {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(function() {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("builder.form()", function() {
  it("renders a form tag", function() {
    act(function() {
      render(<builder.form></builder.form>, container);
    });

    expect(container.children[0].tagName).toEqual("FORM");
  });

  it("forwards the ref", function() {
    const testRef = React.createRef(null);

    act(function() {
      render(<builder.form ref={testRef}></builder.form>, container);
    });

    expect(container.children[0]).toEqual(testRef.current);
  });

  it("renders the children between the form tag", function() {
    act(function() {
      render(
        <builder.form>
          <div>Test immediate children</div>
          <div>
            <span>Test inner children</span>
          </div>
        </builder.form>, container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});

describe("builder.input()", function() {
  it("renders an input tag", function() {
    act(function() {
      render(<builder.input />, container);
    });

    expect(container.children[0].tagName).toEqual("INPUT");
  });

  it("forwards the ref", function() {
    const testRef = React.createRef(null);

    act(function() {
      render(<builder.input ref={testRef} />, container);
    });

    expect(container.children[0]).toEqual(testRef.current);
  });
});

describe("builder.select()", function() {
  it("renders a select tag", function() {
    act(function() {
      render(<builder.select></builder.select>, container);
    });

    expect(container.children[0].tagName).toEqual("SELECT");
  });

  it("forwards the ref", function() {
    const testRef = React.createRef(null);

    act(function() {
      render(<builder.select ref={testRef}></builder.select>, container);
    });

    expect(container.children[0]).toEqual(testRef.current);
  });

  it("renders the children between the form tag", function() {
    act(function() {
      render(
        <builder.select>
          <option value="test-value-one">Test option one</option>
          <option value="test-value-two">Test option two</option>
          <option value="test-value-any">Test option any</option>
        </builder.select>, container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});

describe("builder.textarea()", function() {
  it("renders a textarea tag", function() {
    act(function() {
      render(<builder.textarea />, container);
    });

    expect(container.children[0].tagName).toEqual("TEXTAREA");
  });

  it("forwards the ref", function() {
    const testRef = React.createRef(null);

    act(function() {
      render(<builder.textarea ref={testRef} />, container);
    });

    expect(container.children[0]).toEqual(testRef.current);
  });
});

describe("form builder", function() {
  it("sets derived name for input element", function() {
    act(function() {
      render(
        <builder.form name="test">
          <builder.input name="input" />
          <builder.textarea name="textarea" />
          <builder.select name="select">
            <option value="test-value-one">Test option one</option>
            <option value="test-value-two">Test option two</option>
            <option value="test-value-any">Test option any</option>
          </builder.select>
        </builder.form>, container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
