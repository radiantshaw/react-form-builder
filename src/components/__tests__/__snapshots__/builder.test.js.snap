// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder.fields() renders the children directly 1`] = `
"<div>Test immediate children</div>
<div><span>Test inner children</span></div>"
`;

exports[`builder.form() renders the children between the opening and closing form tags 1`] = `
"<form>
  <div>Test immediate children</div>
  <div><span>Test inner children</span></div>
</form>"
`;

exports[`builder.select() renders the children between the opening and closing select tags 1`] = `
"<select>
  <option value=\\"test-value-one\\">Test option one</option>
  <option value=\\"test-value-two\\">Test option two</option>
  <option value=\\"test-value-any\\">Test option any</option>
</select>"
`;

exports[`form builder can derive name for a collection as well for the root form 1`] = `
"<form><input name=\\"test[][input]\\"><textarea name=\\"test[][textarea]\\"></textarea><select name=\\"test[][select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;

exports[`form builder can derive name for a collection as well using fields builder 1`] = `
"<form><input name=\\"test[fields][][input]\\"><textarea name=\\"test[fields][][textarea]\\"></textarea><select name=\\"test[fields][][select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;

exports[`form builder can derive name for a collection for individual fields 1`] = `
"<input name=\\"test[fields][input][]\\"><textarea name=\\"test[fields][textarea][]\\"></textarea><select name=\\"test[fields][select][]\\">
  <option value=\\"test-value-one\\">Test option one</option>
  <option value=\\"test-value-two\\">Test option two</option>
  <option value=\\"test-value-any\\">Test option any</option>
</select>"
`;

exports[`form builder can set the derived name for form inputs without a root form 1`] = `
"<input name=\\"test[fields][input]\\"><textarea name=\\"test[fields][textarea]\\"></textarea><select name=\\"test[fields][select]\\">
  <option value=\\"test-value-one\\">Test option one</option>
  <option value=\\"test-value-two\\">Test option two</option>
  <option value=\\"test-value-any\\">Test option any</option>
</select>"
`;

exports[`form builder renders input element as it is if builder context is not found 1`] = `"<input name=\\"input\\">"`;

exports[`form builder renders select element as it is if builder context is not found 1`] = `"<select name=\\"select\\"></select>"`;

exports[`form builder renders textarea element as it is if builder context is not found 1`] = `"<textarea name=\\"textarea\\"></textarea>"`;

exports[`form builder sets the derived name for form inputs 1`] = `
"<form><input name=\\"test[input]\\"><textarea name=\\"test[textarea]\\"></textarea><select name=\\"test[select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;

exports[`form builder sets the derived name for form inputs nested one level deep 1`] = `
"<form>
  <div><input name=\\"test[input]\\"><textarea name=\\"test[textarea]\\"></textarea><select name=\\"test[select]\\">
      <option value=\\"test-value-one\\">Test option one</option>
      <option value=\\"test-value-two\\">Test option two</option>
      <option value=\\"test-value-any\\">Test option any</option>
    </select></div>
</form>"
`;

exports[`form builder sets the derived name for form inputs nested several levels deep 1`] = `
"<form>
  <div>
    <div>
      <div><input name=\\"test[input]\\"><textarea name=\\"test[textarea]\\"></textarea><select name=\\"test[select]\\">
          <option value=\\"test-value-one\\">Test option one</option>
          <option value=\\"test-value-two\\">Test option two</option>
          <option value=\\"test-value-any\\">Test option any</option>
        </select></div>
    </div>
  </div>
</form>"
`;

exports[`form builder sets the derived name for form inputs using fields builder 1`] = `
"<form><input name=\\"test[fields][input]\\"><textarea name=\\"test[fields][textarea]\\"></textarea><select name=\\"test[fields][select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;

exports[`form builder sets the derived name for form inputs using proper nested context 1`] = `
"<form><input name=\\"test[pre-outer-input]\\"><textarea name=\\"test[pre-outer-textarea]\\"></textarea><select name=\\"test[pre-outer-select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select><input name=\\"test[fields][inner-input]\\"><textarea name=\\"test[fields][inner-textarea]\\"></textarea><select name=\\"test[fields][inner-select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select><input name=\\"test[post-outer-input]\\"><textarea name=\\"test[post-outer-textarea]\\"></textarea><select name=\\"test[post-outer-select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;

exports[`form builder sets the derived name for form inputs with nested fields 1`] = `
"<form><input name=\\"test[outer-fields][inner-fields][input]\\"><textarea name=\\"test[outer-fields][inner-fields][textarea]\\"></textarea><select name=\\"test[outer-fields][inner-fields][select]\\">
    <option value=\\"test-value-one\\">Test option one</option>
    <option value=\\"test-value-two\\">Test option two</option>
    <option value=\\"test-value-any\\">Test option any</option>
  </select></form>"
`;
